modules/table/src/test/java/org/apache/ignite/internal/table/distributed/index/IndexBuildControllerTest.java

```diff
@@ -0,0 +1,241 @@
+import org.apache.ignite.internal.catalog.descriptors.CatalogIndexDescriptor;
+import org.apache.ignite.internal.catalog.descriptors.CatalogTableDescriptor;
+import org.apache.ignite.internal.configuration.testframework.ConfigurationExtension;
+import org.apache.ignite.internal.configuration.testframework.InjectConfiguration;
+import org.apache.ignite.internal.raft.service.RaftGroupService;
+import org.apache.ignite.internal.replicator.TablePartitionId;
+import org.apache.ignite.internal.schema.configuration.ExtendedTableChange;
+import org.apache.ignite.internal.schema.configuration.TableChange;
+import org.apache.ignite.internal.schema.configuration.TablesConfiguration;
+import org.apache.ignite.internal.schema.configuration.index.HashIndexChange;
+import org.apache.ignite.internal.schema.configuration.index.TableIndexChange;
+import org.apache.ignite.internal.storage.MvPartitionStorage;
+import org.apache.ignite.internal.storage.engine.MvTableStorage;
+import org.apache.ignite.internal.storage.index.IndexStorage;
+import org.apache.ignite.internal.testframework.BaseIgniteAbstractTest;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.mockito.junit.jupiter.MockitoSettings;
+import org.mockito.quality.Strictness;
+
+/** For {@link IndexBuildController} testing. */
+@ExtendWith(ConfigurationExtension.class)
+@ExtendWith(MockitoExtension.class)
+@MockitoSettings(strictness = Strictness.LENIENT)
+public class IndexBuildControllerTest extends BaseIgniteAbstractTest {
+    private static final String TABLE_NAME_0 = "test_table_0";
+
+    private static final String TABLE_NAME_1 = "test_table_1";
+
+    private static final String INDEX_NAME_0 = "test_index_table_0_index_0";
+
+    private static final String INDEX_NAME_1 = "test_index_table_1_index_0";
+
+    @InjectConfiguration
+    private TablesConfiguration tablesConfig;
+
+    @Mock
+    private IndexBuilder indexBuilder;
+
+    @Mock
+    private RaftGroupService raftClient;
+
+    private IndexBuildController controller;
+
+    @BeforeEach
+    void setUp() {
+        CompletableFuture<Void> changeConfigFuture = tablesConfig.change(tablesChange -> {
+            tablesChange.changeTables().create(TABLE_NAME_0, tableChange -> simpleTable(tableChange, 1));
+            tablesChange.changeTables().create(TABLE_NAME_1, tableChange -> simpleTable(tableChange, 2));
+
+            tablesChange.changeIndexes().create(INDEX_NAME_0, indexChange -> simpleIndex(indexChange, 1, 3));
+            tablesChange.changeIndexes().create(INDEX_NAME_1, indexChange -> simpleIndex(indexChange, 2, 5));
+        });
+
+        assertThat(changeConfigFuture, willCompleteSuccessfully());
+
+        controller = new IndexBuildController(indexBuilder, tablesConfig);
+    }
+
+    @Test
+    void testOnIndexCreate() {
```

Comment:
I feel like you check 4 scenarios at once. Can it be split into 4 independent tests then?